### DISCUSSION TITLE
Discussion for Developers

### DISCUSSION THREAD

**Discussion by Special-Niewbie** [YoCodingMonster/OpenFreezeCenter#34] [Ideas]
Hi @YoCodingMonster , I open this, to have between us one direct discussion to keep going with the development.

**Preface:** 
So, I have a lot of wishes and passions because I like the coding world , but in coding I still too fresh and without a background school on it. It doesn't mean that I will asking you here the things because I'm lazy or wherever, just because in meantime I fix the things, it also takes me time as I'm study/ learning in real time. So, please try to understand that time to times I need just a support as friend here and to support your project as well.

Now, let's keep going with the project, I successfully created a `.deb` file for easy life for everyone, and also avoid few problems that the people faced.
I still need to solve some kind of problems and are below:

1) To take out the annoying asking sudo password? How we can do, it seems that even after created the .deb packages - that gives the root privilege - the app still ask the password for each option that you are going to choose.

2) To let the software to run automatically in each every reboot, with the latest fan control chooses (like Normal, Fast , Fastest and so on...). I can successfully run it each every reboot, you can open the Menu by the displayed icon but the software will not start till you will not choose in which Mode you want the Fan to run. It's like fresh run for the system (even if the `conf `file it's created). So, seems that it miss something like `isw@.service` that need to run OpenFreezeCenter like `openfreezecenter@.service` , and into it add the command to run the service that Turn On the FAN mechanism automatically . Which is?

----------------------------------------------------------------------------------------------------------------------------------------
Now let's solve one by one with the first one, I tell you what i have made to solve it, but unfortunately doesn't work and I run out of ideas now.

I have given to all the files bash and python the high privilege with chmod 777 - 775 and 0755.
I have tried to write with special privilege without asking for the pass word, with the command `sudo visudo` and give the following rights under around of 25 line `%sudo` :

```
myusername   ALL=(ALL) NOPASSWD: /usr/local/bin/MSI_OpenFreezeCenter/advanced.py
myusername   ALL=(ALL) NOPASSWD: /usr/local/bin/MSI_OpenFreezeCenter/at_startup.sh
myusername   ALL=(ALL) NOPASSWD: /usr/local/bin/MSI_OpenFreezeCenter/ec_dump.py
myusername   ALL=(ALL) NOPASSWD: /usr/local/bin/MSI_OpenFreezeCenter/indicator.py
myusername   ALL=(ALL) NOPASSWD: /usr/local/bin/MSI_OpenFreezeCenter/install_deps.sh
myusername   ALL=(ALL) NOPASSWD: /usr/local/bin/MSI_OpenFreezeCenter/monitor.py
myusername   ALL=(ALL) NOPASSWD: /usr/local/bin/MSI_OpenFreezeCenter/read_temp_set.py
myusername   ALL=(ALL) NOPASSWD: /usr/local/bin/MSI_OpenFreezeCenter/write_EC.py
```

or even under `root`, like:

```
root   ALL=(ALL) NOPASSWD: /usr/local/bin/MSI_OpenFreezeCenter/advanced.py
root   ALL=(ALL) NOPASSWD: /usr/local/bin/MSI_OpenFreezeCenter/at_startup.sh
root   ALL=(ALL) NOPASSWD: /usr/local/bin/MSI_OpenFreezeCenter/ec_dump.py
root   ALL=(ALL) NOPASSWD: /usr/local/bin/MSI_OpenFreezeCenter/indicator.py
root   ALL=(ALL) NOPASSWD: /usr/local/bin/MSI_OpenFreezeCenter/install_deps.sh
root   ALL=(ALL) NOPASSWD: /usr/local/bin/MSI_OpenFreezeCenter/monitor.py
root   ALL=(ALL) NOPASSWD: /usr/local/bin/MSI_OpenFreezeCenter/read_temp_set.py
root   ALL=(ALL) NOPASSWD: /usr/local/bin/MSI_OpenFreezeCenter/write_EC.py
```

But it seems that doesn't give any effect on that.

Have you some idea?

**Comment by YoCodingMonster** [comment]
Great efforts dube.. it's never late to learn.. and make some friends while doing it😇.

Now for the first problem. I have specified in python files which files need to run with root privilege like `os.system("x-terminal-emulator -e 'bash -c \"sudo nohup python3 ${pkgdir}monitor.py >/dev/null 2>&1\"'")`

I think when you provide the `no password permissions` you need to change this command above and remove sudo from it.

In theory this will remove the root password requirment whenever a file is run. Which will prevent the file from running atall, but since you provided the file with `no password requirment` so this should counter that and run perfectly..

For the second problem, there need to be minor editing in the indicator.py file which will allow the script to apply changes to the EC as soon as it opens. Since you have accomplished to run the app on startup but it dosent apply the fan curve, little tweaking in the indicator.py will do the magic


In the `indicator.py` go to line `255` and there just remove the `sudo` remember to remove the space before `nohup` or the command will not run.

Compile it and test.. if the monitering works without asking for the permissions.. we have made some progress.

**Reply by Special-Niewbie** [reply]
Thanks for your words dude🙏.

Yes, sorry I forgot to write you, I have tried also to take out the sudo command from each python files but it doesn't effect the results, also I believe that the sudo need to keep if I will set the `visudo` file . What make scratch my head it's the visudo that doesn't get any effect...... something else?

**Reply by YoCodingMonster** [reply]
A simple thing if I may suggest is to save the password once the user set it and use it to manage all things afterwards😂.

Just in case nothing works.

The indicator.py script will not pass sudo permissions across itself because appindicator in python dosent work with root permissions. Maybe that's the problem.

**Reply by Special-Niewbie** [reply]
😅 yes I saw that in extreme way could be possible to add into the python line the password (even if it's not the best for security reasons) but if the people wants automated could be one extreme solutions to consider. 

It seems that the NOPASSWRD it doesn't take the effect on the bash files as well....ghhrrrrr 😖.

 Ok, in meantime I'm keeping testing on it, before go to the next point to solve,  how should I write into the python line to include my password?

**Reply by YoCodingMonster** [reply]
try this first

https://askubuntu.com/questions/155791/how-do-i-sudo-a-command-in-a-script-without-being-asked-for-a-password

the things you did for `no password requirment` is good but the code needs to be `os.system('sudo nohup python3 ${pkgdir}monitor.py >/dev/null 2>&1')` insted of `os.system("x-terminal-emulator -e 'bash -c \"sudo nohup python3 ${pkgdir}monitor.py >/dev/null 2>&1\"'")`

**Reply by Special-Niewbie** [reply]
Ok, I have tried, and it seems that it doesn't influence on it, the unique difference it's that it doesn't show the terminal , I have tried with single 'pythone code line here' and with double line as well "python code line here" . Should be some reason of the visudo that doesn't influence the files , I have tried even to write the path as above without the file extensions in `sudoers`. I have tried even to write ALL=(ALL:ALL) NOPASSWD: .......

**Reply by YoCodingMonster** [reply]
hmm let me see into it, I will come up with something tonight (UTC+5:30)

**Reply by Special-Niewbie** [reply]
Thank you man to help 🙏

**Comment by Special-Niewbie** [comment]
Little **update:**
after billions of variations of test this command it's working `myuseraccount   ALL=(ALL:ALL) NOPASSWD:ALL` but I think this gives to the only and `myuseraccount ` the no password to everything and not to the specific file, right? 
If it gives the rights  to everything it means that example `myuseraccount     ALL=(ALL:ALL) NOPASSWD:/usr/local/bin/MSI_OpenFreezeCenter/monitor.py ` it doesn't get the rights without password in this way, but should be something else where `myuseraccount   ALL=(ALL:ALL) NOPASSWD:ALL` can include .......

But example, if i trying to open / double click on `Synaptic Packages Manager `software, the software to open under `myuseraccount `still request me the password, so in this case what this `myuseraccount   ALL=(ALL:ALL) NOPASSWD:ALL` can cover? 
Because if it's not so risk of the security, it's starting to be the solution here....

UPDATE 2:
I starting to have the solution to keep secure to have only under `root `this function and combine with the .deb security file.....

But again strange that the line code to give to an specific file it doesn't work with the NOPASSWD function, could be because it's run under another syntax the service like the try icon (instead of from the file directly ? ) that it doesn't' match with the NOPASSWD to a specific file?

**Reply by YoCodingMonster** [reply]
Can you send me the link to the .deb file you compiled... I will have a look at it to get some idea of what file tree we are working with here

**Comment by Special-Niewbie** [comment]
Here the reupload: https://we.tl/t-3e6ShJndlh

reuploaded with correct icon ....sorry

The MSI Icon will appear in your start menu apps
The app will be in /usr/local/bin/MSI_OpenFreezeCenter/

**Reply by YoCodingMonster** [reply]
Well I will not see the icon in my tray...coz I will extract it on my phone and just have a look at the directories and files....

**Comment by YoCodingMonster** [comment]
replace
`myusername   ALL=(ALL) NOPASSWD: /usr/local/bin/MSI_OpenFreezeCenter/advanced.py
myusername   ALL=(ALL) NOPASSWD: /usr/local/bin/MSI_OpenFreezeCenter/at_startup.sh
myusername   ALL=(ALL) NOPASSWD: /usr/local/bin/MSI_OpenFreezeCenter/ec_dump.py
myusername   ALL=(ALL) NOPASSWD: /usr/local/bin/MSI_OpenFreezeCenter/indicator.py
myusername   ALL=(ALL) NOPASSWD: /usr/local/bin/MSI_OpenFreezeCenter/install_deps.sh
myusername   ALL=(ALL) NOPASSWD: /usr/local/bin/MSI_OpenFreezeCenter/monitor.py
myusername   ALL=(ALL) NOPASSWD: /usr/local/bin/MSI_OpenFreezeCenter/read_temp_set.py
myusername   ALL=(ALL) NOPASSWD: /usr/local/bin/MSI_OpenFreezeCenter/write_EC.py`
and
`root   ALL=(ALL) NOPASSWD: /usr/local/bin/MSI_OpenFreezeCenter/advanced.py
root   ALL=(ALL) NOPASSWD: /usr/local/bin/MSI_OpenFreezeCenter/at_startup.sh
root   ALL=(ALL) NOPASSWD: /usr/local/bin/MSI_OpenFreezeCenter/ec_dump.py
root   ALL=(ALL) NOPASSWD: /usr/local/bin/MSI_OpenFreezeCenter/indicator.py
root   ALL=(ALL) NOPASSWD: /usr/local/bin/MSI_OpenFreezeCenter/install_deps.sh
root   ALL=(ALL) NOPASSWD: /usr/local/bin/MSI_OpenFreezeCenter/monitor.py
root   ALL=(ALL) NOPASSWD: /usr/local/bin/MSI_OpenFreezeCenter/read_temp_set.py
root   ALL=(ALL) NOPASSWD: /usr/local/bin/MSI_OpenFreezeCenter/write_EC.py`
with
`myusername ALL=(root) NOPASSWD:/usr/bin/python3`
below the `%sudo` comand

the above command in place of your all commands will do the trick.
Here the python3 command gets the upper hand and anything containing the `python3` in command will not ask for the sudo password.
`/usr/bin/python3` is the location of the python3 command and you can get your specific location by opening a terminal and typing `which python3`. this will give you your command install location and this problem is solved!!

**Reply by Special-Niewbie** [reply]
Unfortunately it doesn't work.... I believe should be some specific syntax or combination of that - that should unlock for this specific scenario

**Reply by YoCodingMonster** [reply]
replace `install_deps.sh` file with the file in link below
[https://drive.google.com/file/d/1xWJ2GEPb2YEqKMRr_5kYCw2TWcyKigyj/view?usp=sharing](url)

remove the `conf.txt` from `the location where the debian package installed it` and reinstall the software.

that will work 100%

**Reply by Special-Niewbie** [reply]
I will let you know in some hour 😉, because now I have to go.....keep in touch later

**Comment by Special-Niewbie** [comment]
Ok @YoCodingMonster , I explain you what it's going on with those settings and why I changed few syntax to fix some problem and why I changed them (as it needs to be run in different Distros) I'm trying to list you out point by point  where are the problems:
**1)**
```
	sudo install -Dm 644 etc/modprobe.d/ec_sys.conf $(pwd)/etc/modprobe.d/ec_sys.conf
	sudo install -Dm 644 etc/modules-load.d/ec_sys.conf $(pwd)/etc/modules-load.d/ec_sys.conf
```
These syntax code are wrong , I don't know why, maybe in our days the new Distros menega the sudo install in another waybut the correct syntax  are as below (it's for this reason why the people cannot run the app because the code goes in error and doesn't copy the related file into the related folders, this it's the Fix:
```
           sudo install etc/modprobe.d/ec_sys.conf /etc/modprobe.d/ec_sys.conf
	   sudo install etc/modules-load.d/ec_sys.conf /etc/modules-load.d/ec_sys.conf
```
I spent a lot of time with test and to discover above fix. Also, the 644 "user" seems it's not suitable to give the rights to copy these files.

**2)** 

`sudo modprobe ec_sys write_support=1`

**it's ok**

**3)**

`sudo apt-get install python3.8 -y`

All the Distros, in our days, it uses the Python3 with Python3.9 ; so what happen when you force to install the Python3.8 , it happen that the software it will not use it because will use the Python3 /3.9 because almost the people use the Python basica and upgradesd that comes with the own Distro (except who is a coder). I tell you this because I already experienced with your software about and I have got this experience from other protects like Piper. Piper it's programmed with an specific Python version and if you don't have this specific version the software will freeze and not work. And to let it work you have to prepare the Enviroment without make a mess with your Distro or force Piper to use one of the Python version  that you have installed in your system. But to whom doesn't know it they will say that doesn't work. So, my suggestion it's to not force the people to install the 3.8 because we will face many end user problems and tell them that they have to set the `export PATH `and so on.

**4)**

`sudo apt install python3-pip - y`

**OK**

**5)**
`sudo apt-get install python3.8-tk -y`

The same story like Python3.8 here should be `sudo apt-get install python3-tk -y`

**6)**

`sudo pip3 install psutil -y`

**OK**

**7)**

```
sudo apt install nautilus -y
sudo apt install nautilus-admin -y
```
Nautilus, if you have a Distro like the one that I'm using that doesn't' have Nautilus as default File Manager, if the end user will not uninstall the default File Manager the software will never use and he will find a lot of software as useless (especially if to the end user likes to use another File Manager) and even in Ubuntu it doesn't influence to overpass the `sudo password `and even it doesn't copy the `ec_sys.conf`. For this reason, the .deb package resolve this problem.

**8**
`sudo apt-get install gir1.2-appindicator3-0.1 -y`

**OK**

**9)**

`sudo GRUB_CMDLINE_LINUX_DEFAULT="quiet ec_sys.write_support=1"`

this command should add this voice into `sudo gedit /etc/default/grub`, right? Because I use it, for the Pass Trough GPU on VM enviroment, I watched into default Grub file but unfortunately just then i noticed that it doesn't add the `quiet ec_sys.write_support=1` . So, instead of use the Service, you are using the Grub pre-boot to run the latest Fun settings after Reboot the System, right?

**10)**
```
path_python3=$(which python3)
    user=$(whoami)
    line="$user ALL=(root) NOPASSWD:$path_python3"
    file='/etc/sudoers'
    line_num=$(sudo awk '/%sudo/{print NR}' $file)
    line_num=`expr $line_num + 1` 
    sudo sed -i "$line_num i $line" $file
```
this coding working well and add the voice into sudoers.

With the .DEB packages seems it's better because keep the rights to enter better into the system , now it's set to 777 but later I will decrease the user and with the root in this specifi circumstance the software will not find any obstacle of read and write as can meet if the software it's running with the free files ;)
So, in this way it doesn't solve but even gives the errors in therminal with some limitation of write the files to the related software and to the GRUB.

PS: The test I made just then with all the installations (even Nautilus and python3.8 and so on)

**Reply by YoCodingMonster** [reply]
Ok those are some good changes...
What's the final result after these changes?

Does the sudo problem goes away?

And the problem regarding not getting a installed on other distros is that the install_deps.sh file is not written very effeciently by me...Those changes you did seems to be great!!

**Reply by Special-Niewbie** [reply]
Yes the people face because the install_deps  .sh it doesn't perform , and will perform perfectly with the changes that i made, and it's working with different Distros. For this reason the people complain that doesn't work , because they miss some vital file to let it work.....
Now .deb file will fix few problems about the chmod , automatically. And if there is no one specific fix for the Sudo Password I suggest to use :

`myusername ALL=(root) NOPASSWD:ALL`

Because it's more close `Group` as this Group can only access the specific `.deb ` file system that as got this particular Software permissions of` myusername` combine with `root` so for security reason it's far away much better than myusername ALL=(ALL) NOPASSWD:ALL or even sudo ALL=(ALL) NOPASSWD:ALL.

**Comment by Special-Niewbie** [comment]
So the point `1)` for now we can say that it's close story (if you will have some other idea I will be happy to experiment).
Lets go to the point `2)` :

`To let the software to run automatically in each every reboot, with the latest fan control chooses (like Normal, Fast , Fastest and so on...). I can successfully run it each every reboot, you can open the Menu by the displayed icon but the software will not start till you will not choose in which Mode you want the Fan to run. It's like fresh run for the system (even if the conf file it's created). So, seems that it miss something like isw@.service that need to run OpenFreezeCenter like openfreezecenter@.service , and into it add the command to run the service that Turn On the FAN mechanism automatically . Which is?`

What we should do here? Should be created a Service file or I have to try to write into the GRUB this: `"quiet ec_sys.write_support=1"` ?

**Reply by YoCodingMonster** [reply]
No we don't need to add any services and all.. I just need to script in python that at fresh start always apply the fan mode which is saved in conf.txt file...that's easy and I will send you the modified indicator.py by tommorow.

**Reply by Special-Niewbie** [reply]
Love you man, number one ;) .....

**Reply by YoCodingMonster** [reply]
https://drive.google.com/file/d/1StSztisPhGNGR0sjMhFgLLpYtDDYXDQq/view?usp=sharing

That is the link t the new indicator.py file. replace the old file with this and on every startup, the fan profile should be applied!!

**Reply by YoCodingMonster** [reply]
> https://drive.google.com/file/d/1StSztisPhGNGR0sjMhFgLLpYtDDYXDQq/view?usp=sharing
> 
> That is the link t the new indicator.py file. replace the old file with this and on every startup, the fan profile should be applied!!

Does it work?

**Reply by Special-Niewbie** [reply]
> > https://drive.google.com/file/d/1StSztisPhGNGR0sjMhFgLLpYtDDYXDQq/view?usp=sharing
> > 
> > That is the link t the new indicator.py file. replace the old file with this and on every startup, the fan profile should be applied!!
> 
> Does it work?

Morning @YoCodingMonster , I just see now your message, yesterday was late in my side . Let me test test and I will report you back as soon as possible....

**Reply by YoCodingMonster** [reply]
> > > https://drive.google.com/file/d/1StSztisPhGNGR0sjMhFgLLpYtDDYXDQq/view?usp=sharing
> > > 
> > > That is the link t the new indicator.py file. replace the old file with this and on every startup, the fan profile should be applied!!
> > 
> > Does it work?
> 
> Morning @YoCodingMonster , I just see now your message, yesterday was late in my side . Let me test test and I will report you back as soon as possible....

What time it is there..
At my place it's 9:45 AM

**Reply by Special-Niewbie** [reply]
> > > > https://drive.google.com/file/d/1StSztisPhGNGR0sjMhFgLLpYtDDYXDQq/view?usp=sharing
> > > > 
> > > > That is the link t the new indicator.py file. replace the old file with this and on every startup, the fan profile should be applied!!
> > > 
> > > Does it work?
> > 
> > Morning @YoCodingMonster , I just see now your message, yesterday was late in my side . Let me test test and I will report you back as soon as possible....
> 
> What time it is there..
> At my place it's 9:45 AM

Because I'm living in Asia not Europe 😉

**Reply by YoCodingMonster** [reply]
> > > > > https://drive.google.com/file/d/1StSztisPhGNGR0sjMhFgLLpYtDDYXDQq/view?usp=sharing
> > > > > 
> > > > > That is the link t the new indicator.py file. replace the old file with this and on every startup, the fan profile should be applied!!
> > > > 
> > > > Does it work?
> > > 
> > > Morning @YoCodingMonster , I just see now your message, yesterday was late in my side . Let me test test and I will report you back as soon as possible....
> > 
> > What time it is there..
> > At my place it's 9:45 AM
> 
> Because I'm living in Asia not Europe 😉

I am in Asia too....India to be specific!

**Reply by Special-Niewbie** [reply]
In Shenzhen Sud of China 12.30

**Comment by Special-Niewbie** [comment]
Congrats Man, it's working :+1:

**Reply by Special-Niewbie** [reply]
Now, I will do all the corrections from zero ,delete all the system correction that I did by hands and repack. And let's see if everything will work in fresh way

**Reply by YoCodingMonster** [reply]
> Now, I will do all the corrections from zero ,delete all the system correction that I did by hands and repack. And let's see if everything will work in fresh way 

Can you share the bed package with me after its done!!

**Reply by Special-Niewbie** [reply]
This for sure ;)....

**Comment by Special-Niewbie** [comment]
Hi @YoCodingMonster , ok I fixed some problems with the packages and finalized everything , now the package can install and uninstall and working well. But still last thing I don't know for which reason the `at_startup.sh` it jumps to run `install_deps.sh` (of course if the `conf.txt`, there isn't into the folder). 
I gave all the permissions and executions but when I launch the `at_startup.sh` it starts only the indicator.py, and it's a problem this because will not set well the software as it miss all the dependencies from `install_deps.sh` ( `install_deps.sh `can run only if you execute by it self and then use always the `at_startup.sh` ) . I have tried some fix but it doesn't want to run in any way (except manual), do you have some other idea how we can let it run for the first time automatically, when it start the `at_startup.sh` ?

**Reply by YoCodingMonster** [reply]
I think we need to change what files we check for the dependencies to get installed and make the location of that to something outside the debian package itself!!.  I will give the edited file to you in sometime today's!!

**Reply by Special-Niewbie** [reply]
> I think we need to change what files we check for the dependencies to get installed and make the location of that to something outside the debian package itself!!.  I will give the edited file to you in sometime today's!!

I can create the Deb to tell to put the install_deps.sh in the Desktop and then the user will install , but would be strange. Or I can create one forwarded that will add the packages to execute only the install_dep.sh (named like Open Freeze Center dependencies) . Or I was thinking to add all the list of the dependencies into the python indicator that will run those things only if miss the conf.txt file. The same job as make the install_deps.sh. But what cannot understand, it's why the two bash cannot communicate...

**Reply by YoCodingMonster** [reply]
> > I think we need to change what files we check for the dependencies to get installed and make the location of that to something outside the debian package itself!!.  I will give the edited file to you in sometime today's!!
> 
> I can create the Deb to tell to put the install_deps.sh in the Desktop and then the user will install , but would be strange. Or I can create one forwarded that will add the packages to execute only the install_dep.sh (named like Open Freeze Center dependencies) . Or I was thinking to add all the list of the dependencies into the python indicator that will run those things only if miss the conf.txt file. The same job as make the install_deps.sh. But what cannot understand, it's why the two bash cannot communicate...

May I have a look at the deb package, and the package before build.

Let me see the structure!

**Comment by Special-Niewbie** [comment]
@YoCodingMonster , after many tests and differentiation of the coding maybe I got why the `install_deps.sh` it's jumped by the `at_startup.sh`... Do you know the exactly function of the bash to write in between `./install_deps.sh` and `python3 indicator.py >/dev/null 2>&1 & exit` into `at_startup.sh` ?
I need that `at_startup.sh` starts first `./install_deps.sh` and after 1 or even 2 seconds will start `python3 indicator.py >/dev/null 2>&1 & exit` function, now because this processes to run these two functions are so fast, the `install_deps.sh` it's like that read every time that there is the conf.txt into the folder.... For this reason it doesn't run all the bash file

**Reply by YoCodingMonster** [reply]
> @YoCodingMonster , after many tests and differentiation of the coding maybe I got why the `install_deps.sh` it's jumped by the `at_startup.sh`... Do you know the exactly function of the bash to write in between `./install_deps.sh` and `python3 indicator.py >/dev/null 2>&1 & exit` into `at_startup.sh` ?
> I need that `at_startup.sh` starts first `./install_deps.sh` and after 1 or even 2 seconds will start `python3 indicator.py >/dev/null 2>&1 & exit` function, now because this processes to run these two functions are so fast, the `install_deps.sh` it's like that read every time that there is the conf.txt into the folder.... For this reason it doesn't run all the bash file

Yeah I can do that. I know the code..will make the pauses as said.

**Comment by Special-Niewbie** [comment]
Excuse me @YoCodingMonster , if text you many times but on so many test that I make i need to centralized where is the exactly problem and finally I got it. Please, be patient so I tell you the story here.
I have tried many thing to the bash like `set -x ` change the coding into the bash delete `IF` functions and so on so on so on.....
The problem it's the forwarding `.desktop` icon I have into it the following code:
```

[Desktop Entry]
Name=MSI OpenFreezeCenter
Exec=/usr/local/bin/MSI_OpenFreezeCenter/at_startup.sh
Type=Application
StartupNotify=true
Icon=/usr/local/bin/MSI_OpenFreezeCenter/icon_22.png
Path=/usr/local/bin/MSI_OpenFreezeCenter/
```
So what happen, it seems that the `.desktop` (Desktop Entry icon) if set the forwarding to run the `startup.sh ` for some reason it escape the `install_deps.sh` even if it's included into the `at_startup.sh`, so if I add/ invoke in the list the `Terminal=true` function into the `.desktop` forwarder it runs the `install_deps.sh` and shows you the Terminal that tells you to add the sudo password to continue to proceed with `install_deps.sh`  but will not run the app `indicator.py` (so reverse story). The `install_deps.sh` it needs the sudo password in terminal to run but maybe could be this the reason why it doesn't run, because there is not any terminal pop-up from forwarder as it's set without Terminal=true and it's kind of a forwarder set for run the app pre-compiled like python `indicator.py` pass trough by the bash; but it's not like to run it by the normal Terminal app when you install manually with `./install_deps.sh`. I believe it's a bit different this slightly differentiation, now to let coexist together the two different bash it seems there is not a clear road here, and  the unique solution it's to include the all `IF` function with all the dependencies from `install_deps.sh` into the `indicator.py` or to delete the `install_deps.sh` create a separate file in .py named `install_deps.py` that I can run as an app like the `indicator.py` from the `.desktop` (Desktop Entry icon) by the script `at_startup.sh` and change the coding into `at_startup.sh` as following:

```
#!/bin/sh
python3 install_deps.py 
python3 indicator.py >/dev/null 2>&1 & exit
```

Please , when you will have time try to create this new file and finally we have solved everything in this way ;)

**Reply by YoCodingMonster** [reply]
Sure I will work on this file as soon as I finish my project from college

**Reply by Special-Niewbie** [reply]
> Sure I will work on this file as soon as I finish my project from college

Take your time and good luck with your project 😉. Thank you very much

**Reply by YoCodingMonster** [reply]
done some changes to the files.
replace the

indicator.py
at_startup.sh
install_deps.sh

with new versions i provide in zip file. -> https://drive.google.com/file/d/1PIzJnaErM0A-9j0BylmWKXqK1KksZaqa/view?usp=sharing

there is also a new file to write the sudoers calles root_permission.sh which need to be placed in the same directory as all other files.

**Reply by Special-Niewbie** [reply]
Done!!!!!!! Ok, I finished to test and fixing the cooding because in this way as you made was some problem. Anyway, it's exactly what i mean, soon I will push you in GitHub all the changes to do to let read well the EC from the beginning and make all the automated installation from the beginning. Congratulations Man, this app start to be a real app now........Later we will fix the minor problems like the DPI, Double GPU showing on Monitoring (IF will be 2 of them) and whatever else it need. 
THANK YOU SO SO SO SO SO MUCH!!! :+1:

**Reply by YoCodingMonster** [reply]
dont push your forked line into mine, i quite dont have a lot of experience merging it.. just send me a link to the .deb package. the files i have already updated in the main repository!

or just send me a merge request, i will do it properly this time. i have screwed a merge in past XD

**Reply by Special-Niewbie** [reply]
Ok, here are the files for this New version 1.2 , now the people will not complain anymore about the installation ;).

To install the .deb package will be installed in normal way and to uninstall the .deb package it needs the following command in the terminal: `sudo dpkg --remove msi-openfreezecenter `

This will uninstall anything that the package was created during the installation process; but not the `conf.txt` file for now, this because the `conf.txt `file it's created after the package was created the system folder (but really it doesn't matter for the final user).

All the Software files, are and should be found into `/usr/local/bin/MSI_OpenFreezeCenter/`
I have created the shortcut that the people will find into them Start List software, and to run the software automatically they need just to select the shortcut and give the Auto-start in each every reboot from them OS Settings.

These are the files: https://we.tl/t-2LHeOBg7Dm

In the beginning, the software will ask the password for first accesses, but after nothing and if you need I will tell you also why I made the changes and why they works ;)

**Reply by YoCodingMonster** [reply]
now I have some questions.

1. the install_deps.sh contain the line to edit sudoers, which creates a problem when getting the user name for sudoers and add a line starting with root in sudoers instead of the user name.
2. instead of `line="$user ALL=(root) NOPASSWD:$path_python3"` you replaced it with `line="$user ALL=(root) NOPASSWD:ALL` which means the root can perform any command in the system without root passowrd.

thats it, let me know the benefits of using these edits!

**Reply by Special-Niewbie** [reply]
Sure , the `line="$user ALL=(root) NOPASSWD:$path_python3"` I don't know for which reason it doesn't work at all (and should works...) these past days I have tested any variation of that to let it work, I have even consulted some Linux coder and tried to fill the `Cmnd alias specification ` as well, but nothing it works. Probably because before the python3 and the .py file if you launch the command (like example Monitoring or whatever from the Icon-tray menu) there is some extra input command from the system icon tray that doesn't let work the `NOPASSWD:$path_python3 `, I didn't discover yet which is the exactly the extra command that "cheat" the `line="$user ALL=(root) NOPASSWD:$path_python3"` to don't work where the `ALL `command it's include, this extra step in back ground.....

Otherwise will not work the `NOPASSWD:` function (even if the sudoers it's filled). I was thinking to use some Linux OS Debugging to understand what it jumps to don't let it work the `line="$user ALL=(root) NOPASSWD:$path_python3"` ....

**Reply by Special-Niewbie** [reply]
the first point sorry I have jumped ....

The differences it's that, instead of creates the `root ALL=(root) NOPASSWD:ALL`  unfortunately the software it doesn't work and it creates the problems; because when you call with the app for the first time the Python , the command will start always with your User name/group first and then enter with the root name user/group, so if it's not `"user" ALL=(root) NOPASSWD:ALL` the software cannot overpass the first terminal command and ask you continually the passwords , for this reason it needs to change in user and than in a specific group as root.

**Reply by Special-Niewbie** [reply]
Also the root.sh I have to delete, because as you made in each every boot the file will create a new same line called `ROOT ALL=(root) NOPASSWD:$path_python3 `, and happens that the `sudoers `file will be filled with a long list of the same USER special `NOPASSWD`, instead with `if` commnad that it will create one time (the first time as need it) and that's it, without duplicate infinitely the name into `sudoers` file.

**Reply by Special-Niewbie** [reply]
I think to solve the `sudoers` for **NOPASSWD** function and gives more specific function to protect the privacy it's not `user ALL=(root) NOPASSWD:$path_python3` but it's to find the way to add the specific command opened in Terminal in `"x-terminal-emulator -e 'bash -c \"sudo nohup python3 ${pkgdir}monitor.py >/dev/null 2>&1\"'"` or something similar string. Because the problem of the password it's not the actual Python file or Python Software to get the privilege , but it's the actual Terminal string function that it needs the `NOPASSWD` to the specific function of the app with Terminal (like Monitoring and so on), otherwise I found that it doesn't work and keep you asking the Password. I hope to discover it.....

**Reply by YoCodingMonster** [reply]
> I think to solve the `sudoers` for **NOPASSWD** function and gives more specific function to protect the privacy it's not `user ALL=(root) NOPASSWD:$path_python3` but it's to find the way to add the specific command opened in Terminal in `sudoers` , like : `x-terminal-emulator -e 'bash -c \"sudo ./install_deps.sh\"` or something similar string. Because the problem of the password it's not the actual Python file or Python Software to get the privilege , but it's the actual Terminal string function that it needs the `NOPASSWD` to the specific function of the app with Terminal (like Monitoring and so on), otherwise I found that it doesn't work and keep you asking the Password. I hope to discover it.....

Hmm I am working on it..and adding some more functionality to it

**Reply by Special-Niewbie** [reply]
Sorry I wanna say like something like this into `sudoers`: `"x-terminal-emulator -e 'bash -c \"sudo nohup python3 ${pkgdir}monitor.py >/dev/null 2>&1\"'"`......

In sudoers we can only provide the function name... Not the complete combination of functions we use..neither it works for the whole statements.....

I was thinking to create a custom function to use in sudoers

**Reply by YoCodingMonster** [reply]
I will be creating a custom command this week and let you know. What's the progress!!
